File: player-core/src/main/java/com/libreshockwave/player/debug/DebugController.java
package com.libreshockwave.player.debug;

import com.libreshockwave.chunks.ScriptChunk;
import com.libreshockwave.vm.Datum;
import com.libreshockwave.vm.TraceListener;

import javax.swing.*;
import java.util.*;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.CopyOnWriteArrayList;
import java.util.concurrent.Semaphore;

/**
 * Core debugging controller that implements TraceListener.
 * Blocks VM execution when paused by using a Semaphore in onInstruction().
 */
public class DebugController implements TraceListener {

    /**
     * Current debug state.
     */
    public enum DebugState {
        /** VM is running normally */
        RUNNING,
        /** VM is paused at a breakpoint or step */
        PAUSED,
        /** VM is executing a step command */
        STEPPING
    }

    /**
     * Step mode for single-stepping.
     */
    public enum StepMode {
        /** No stepping - run normally */
        NONE,
        /** Break on every instruction */
        STEP_INTO,
        /** Break when returning to same or lower call depth */
        STEP_OVER,
        /** Break when returning to lower call depth */
        STEP_OUT
    }

    /**
     * Represents a single frame in the call stack.
     */
    public record CallFrame(
        int scriptId,
        String scriptName,
        String handlerName,
        List<Datum> arguments,
        Datum receiver
    ) {}

    // Synchronization - blocks VM thread when paused
    private final Semaphore pauseSemaphore = new Semaphore(0);
    private final Object stateLock = new Object();

    // State
    private volatile DebugState state = DebugState.RUNNING;
    private volatile StepMode stepMode = StepMode.NONE;
    private volatile int callDepth = 0;
    private volatile int targetCallDepth = 0;

    // Request to pause on next instruction (when Pause button is clicked)
    private volatile boolean pauseRequested = false;

    // Breakpoints: scriptId -> set of instruction offsets
    private final Map<Integer, Set<Integer>> breakpoints = new ConcurrentHashMap<>();

    // Current handler context (for snapshot creation)
    private volatile HandlerInfo currentHandlerInfo;
    private volatile ScriptChunk currentScript;
    private volatile ScriptChunk.Handler currentHandler;
    private volatile InstructionInfo currentInstructionInfo;

    // Call stack tracking
    private final List<CallFrame> callStack = new ArrayList<>();

    // Most recent snapshot (for UI)
    private volatile DebugSnapshot currentSnapshot;

    // UI listeners (notified on EDT)
    private final List<DebugStateListener> listeners = new CopyOnWriteArrayList<>();

    // Globals accessor (set by Player)
    private volatile Map<String, Datum> globalsSnapshot = Collections.emptyMap();

    // Delegate trace listener (for UI updates)
    private volatile TraceListener delegateListener;

    public DebugController() {
    }

    /**
     * Set a delegate trace listener that receives all events.
     * This allows the UI to observe VM execution while the controller handles debugging.
     */
    public void setDelegateListener(TraceListener listener) {
        this.delegateListener = listener;
    }

    // TraceListener implementation

    @Override
    public void onHandlerEnter(HandlerInfo info) {
        callDepth++;
        currentHandlerInfo = info;

        // Track call stack
        synchronized (callStack) {
            callStack.add(new CallFrame(
                info.scriptId(),
                info.scriptDisplayName(),
                info.handlerName(),
                new ArrayList<>(info.arguments()),
                info.receiver()
            ));
        }

        // Forward to delegate
        TraceListener delegate = delegateListener;
        if (delegate != null) {
            delegate.onHandlerEnter(info);
        }
    }

    @Override
    public void onHandlerExit(HandlerInfo info, Datum returnValue) {
        callDepth--;

        // Pop from call stack
        synchronized (callStack) {
            if (!callStack.isEmpty()) {
                callStack.remove(callStack.size() - 1);
            }
        }

        // Check for step-out completion
        synchronized (stateLock) {
            if (stepMode == StepMode.STEP_OUT && callDepth <= targetCallDepth) {
                // Will pause on next instruction after returning
                stepMode = StepMode.STEP_INTO;
            }
        }

        // Forward to delegate
        TraceListener delegate = delegateListener;
        if (delegate != null) {
            delegate.onHandlerExit(info, returnValue);
        }
    }

    @Override
    public void onInstruction(InstructionInfo info) {
        currentInstructionInfo = info;

        // Forward to delegate
        TraceListener delegate = delegateListener;
        if (delegate != null) {
            delegate.onInstruction(info);
        }

        // Check if we should break
        if (shouldBreak(info)) {
            pauseExecution(info);
        }
    }

    @Override
    public void onVariableSet(String type, String name, Datum value) {
        // Forward to delegate
        TraceListener delegate = delegateListener;
        if (delegate != null) {
            delegate.onVariableSet(type, name, value);
        }
    }

    @Override
    public void onError(String message, Exception error) {
        // Forward to delegate
        TraceListener delegate = delegateListener;
        if (delegate != null) {
            delegate.onError(message, error);
        }
    }

    @Override
    public void onDebugMessage(String message) {
        // Forward to delegate
        TraceListener delegate = delegateListener;
        if (delegate != null) {
            delegate.onDebugMessage(message);
        }
    }

    /**
     * Determine if we should break at this instruction.
     */
    private boolean shouldBreak(InstructionInfo info) {
        synchronized (stateLock) {
            // Check pause request
            if (pauseRequested) {
                pauseRequested = false;
                return true;
            }

            // Check breakpoints
            if (currentHandlerInfo != null && hasBreakpoint(currentHandlerInfo.scriptId(), info.offset())) {
                return true;
            }

            // Check stepping modes
            return switch (stepMode) {
                case STEP_INTO -> true;  // Always break
                case STEP_OVER -> callDepth <= targetCallDepth;
                case STEP_OUT -> false;  // Handled in onHandlerExit
                case NONE -> false;
            };
        }
    }

    /**
     * Pause VM execution at the current instruction.
     * This blocks the VM thread until resume/step is called.
     */
    private void pauseExecution(InstructionInfo info) {
        synchronized (stateLock) {
            state = DebugState.PAUSED;
            stepMode = StepMode.NONE;
        }

        // Capture snapshot
        currentSnapshot = captureSnapshot(info);

        // Notify UI on EDT
        notifyPaused(currentSnapshot);

        // Block VM thread until resumed
        try {
            pauseSemaphore.acquire();
        } catch (InterruptedException e) {
            Thread.currentThread().interrupt();
        }
    }

    /**
     * Capture current state as an immutable snapshot.
     */
    private DebugSnapshot captureSnapshot(InstructionInfo info) {
        HandlerInfo handlerInfo = currentHandlerInfo;
        if (handlerInfo == null) {
            return null;
        }

        // Build instruction list from handler
        List<DebugSnapshot.InstructionDisplay> instructions = new ArrayList<>();
        if (handlerInfo.literals() != null) {
            // We need to get instructions from somewhere
            // The handler info doesn't contain instructions directly
            // We'll build from the current instruction info for now
        }

        // Build locals map from current scope
        // Note: We only have info from HandlerInfo, not direct scope access
        Map<String, Datum> locals = new LinkedHashMap<>();
        // Locals would need to be captured from the scope

        return new DebugSnapshot(
            handlerInfo.scriptId(),
            handlerInfo.scriptDisplayName(),
            handlerInfo.handlerName(),
            info.offset(),
            info.bytecodeIndex(),
            info.opcode(),
            info.argument(),
            info.annotation(),
            instructions,  // Will be populated when we have script access
            new ArrayList<>(info.stackSnapshot()),
            locals,
            new LinkedHashMap<>(globalsSnapshot),
            new ArrayList<>(handlerInfo.arguments()),
            handlerInfo.receiver(),
            getCallStackSnapshot()
        );
    }

    // Control methods (called from UI on EDT)

    /**
     * Step into - execute one instruction and pause.
     */
    public void stepInto() {
        synchronized (stateLock) {
            if (state != DebugState.PAUSED) return;
            stepMode = StepMode.STEP_INTO;
            state = DebugState.STEPPING;
        }
        notifyResumed();
        pauseSemaphore.release();  // Unblock VM thread
    }

    /**
     * Step over - execute until returning to same call depth.
     */
    public void stepOver() {
        synchronized (stateLock) {
            if (state != DebugState.PAUSED) return;
            stepMode = StepMode.STEP_OVER;
            targetCallDepth = callDepth;
            state = DebugState.STEPPING;
        }
        notifyResumed();
        pauseSemaphore.release();
    }

    /**
     * Step out - execute until returning to caller.
     */
    public void stepOut() {
        synchronized (stateLock) {
            if (state != DebugState.PAUSED) return;
            stepMode = StepMode.STEP_OUT;
            targetCallDepth = callDepth - 1;
            state = DebugState.STEPPING;
        }
        notifyResumed();
        pauseSemaphore.release();
    }

    /**
     * Continue execution until next breakpoint.
     */
    public void continueExecution() {
        synchronized (stateLock) {
            if (state != DebugState.PAUSED) return;
            stepMode = StepMode.NONE;
            state = DebugState.RUNNING;
        }
        notifyResumed();
        pauseSemaphore.release();
    }

    /**
     * Request pause on next instruction.
     * Call this when the VM is running to pause it.
     */
    public void pause() {
        synchronized (stateLock) {
            if (state == DebugState.RUNNING || state == DebugState.STEPPING) {
                pauseRequested = true;
            }
        }
    }

    /**
     * Check if currently paused.
     */
    public boolean isPaused() {
        return state == DebugState.PAUSED;
    }

    /**
     * Get current debug state.
     */
    public DebugState getState() {
        return state;
    }

    /**
     * Get current snapshot (may be null if not paused).
     */
    public DebugSnapshot getCurrentSnapshot() {
        return currentSnapshot;
    }

    // Breakpoint management

    /**
     * Toggle a breakpoint at the given script and offset.
     * @return true if breakpoint was added, false if removed
     */
    public boolean toggleBreakpoint(int scriptId, int offset) {
        Set<Integer> offsets = breakpoints.computeIfAbsent(scriptId, k -> ConcurrentHashMap.newKeySet());
        boolean added;
        if (offsets.contains(offset)) {
            offsets.remove(offset);
            added = false;
        } else {
            offsets.add(offset);
            added = true;
        }
        notifyBreakpointsChanged();
        return added;
    }

    /**
     * Add a breakpoint.
     */
    public void addBreakpoint(int scriptId, int offset) {
        breakpoints.computeIfAbsent(scriptId, k -> ConcurrentHashMap.newKeySet()).add(offset);
        notifyBreakpointsChanged();
    }

    /**
     * Remove a breakpoint.
     */
    public void removeBreakpoint(int scriptId, int offset) {
        Set<Integer> offsets = breakpoints.get(scriptId);
        if (offsets != null) {
            offsets.remove(offset);
            notifyBreakpointsChanged();
        }
    }

    /**
     * Check if a breakpoint exists.
     */
    public boolean hasBreakpoint(int scriptId, int offset) {
        Set<Integer> offsets = breakpoints.get(scriptId);
        return offsets != null && offsets.contains(offset);
    }

    /**
     * Get all breakpoints.
     */
    public Map<Integer, Set<Integer>> getBreakpoints() {
        Map<Integer, Set<Integer>> result = new HashMap<>();
        for (Map.Entry<Integer, Set<Integer>> entry : breakpoints.entrySet()) {
            result.put(entry.getKey(), new HashSet<>(entry.getValue()));
        }
        return result;
    }

    /**
     * Clear all breakpoints.
     */
    public void clearAllBreakpoints() {
        breakpoints.clear();
        notifyBreakpointsChanged();
    }

    // Listener management

    /**
     * Add a listener for debug state changes.
     */
    public void addListener(DebugStateListener listener) {
        listeners.add(listener);
    }

    /**
     * Remove a listener.
     */
    public void removeListener(DebugStateListener listener) {
        listeners.remove(listener);
    }

    private void notifyPaused(DebugSnapshot snapshot) {
        SwingUtilities.invokeLater(() -> {
            for (DebugStateListener listener : listeners) {
                listener.onPaused(snapshot);
            }
        });
    }

    private void notifyResumed() {
        SwingUtilities.invokeLater(() -> {
            for (DebugStateListener listener : listeners) {
                listener.onResumed();
            }
        });
    }

    private void notifyBreakpointsChanged() {
        SwingUtilities.invokeLater(() -> {
            for (DebugStateListener listener : listeners) {
                listener.onBreakpointsChanged();
            }
        });
    }

    // State injection (called by Player)

    /**
     * Update the globals snapshot for debugging.
     * Called by Player before/during execution.
     */
    public void setGlobalsSnapshot(Map<String, Datum> globals) {
        this.globalsSnapshot = globals != null ? new LinkedHashMap<>(globals) : Collections.emptyMap();
    }

    /**
     * Reset debug state (call when movie is stopped/reset).
     */
    public void reset() {
        synchronized (stateLock) {
            // If paused, release the semaphore to avoid deadlock
            if (state == DebugState.PAUSED) {
                state = DebugState.RUNNING;
                pauseSemaphore.release();
            }
            state = DebugState.RUNNING;
            stepMode = StepMode.NONE;
            callDepth = 0;
            targetCallDepth = 0;
            pauseRequested = false;
        }
        // Drain any stale semaphore permits to avoid unexpected behavior
        pauseSemaphore.drainPermits();
        synchronized (callStack) {
            callStack.clear();
        }
        currentHandlerInfo = null;
        currentInstructionInfo = null;
        currentSnapshot = null;
    }

    /**
     * Get current call depth.
     */
    public int getCallDepth() {
        return callDepth;
    }

    /**
     * Get current handler name (for display).
     */
    public String getCurrentHandlerName() {
        HandlerInfo info = currentHandlerInfo;
        return info != null ? info.handlerName() : null;
    }

    /**
     * Get a snapshot of the current call stack.
     * Returns a copy to avoid concurrency issues.
     */
    public List<CallFrame> getCallStackSnapshot() {
        synchronized (callStack) {
            return new ArrayList<>(callStack);
        }
    }
}

File: cpp/src/player/debug/DebugController.cpp
#include "libreshockwave/player/debug/DebugController.hpp"

#include <algorithm>
#include <utility>

namespace libreshockwave::player::debug {

void DebugController::setDelegateListener(std::shared_ptr<lingo::vm::TraceListener> listener) {
    std::lock_guard lock(mutex_);
    delegateListener_ = std::move(listener);
}

void DebugController::addListener(DebugStateListener* listener) {
    if (listener == nullptr) {
        return;
    }
    std::lock_guard lock(mutex_);
    if (std::find(listeners_.begin(), listeners_.end(), listener) == listeners_.end()) {
        listeners_.push_back(listener);
    }
}

void DebugController::removeListener(DebugStateListener* listener) {
    std::lock_guard lock(mutex_);
    listeners_.erase(std::remove(listeners_.begin(), listeners_.end(), listener), listeners_.end());
}

void DebugController::onHandlerEnter(const HandlerInfo& info) {
    std::shared_ptr<lingo::vm::TraceListener> delegate;
    {
        std::lock_guard lock(mutex_);
        ++callDepth_;
        currentHandlerInfo_ = info;
        handlerInfoStack_.push_back(info);
        callStack_.push_back(CallFrame{
            info.scriptId,
            info.scriptDisplayName,
            info.handlerName,
            info.arguments,
            info.receiver.isVoid() ? std::optional<lingo::Datum>{} : std::optional<lingo::Datum>{info.receiver},
        });
        delegate = delegateListener_;
    }
    if (delegate) {
        delegate->onHandlerEnter(info);
    }
}

void DebugController::onHandlerExit(const HandlerInfo& info, const lingo::Datum& returnValue) {
    std::shared_ptr<lingo::vm::TraceListener> delegate;
    {
        std::lock_guard lock(mutex_);
        if (callDepth_ > 0) {
            --callDepth_;
        }
        if (!handlerInfoStack_.empty()) {
            handlerInfoStack_.pop_back();
        }
        currentHandlerInfo_ = handlerInfoStack_.empty()
            ? std::nullopt
            : std::optional<HandlerInfo>{handlerInfoStack_.back()};
        if (!callStack_.empty()) {
            callStack_.pop_back();
        }
        if (stepMode_ == StepMode::StepOut && callDepth_ <= targetCallDepth_) {
            stepMode_ = StepMode::StepInto;
        }
        delegate = delegateListener_;
    }
    if (delegate) {
        delegate->onHandlerExit(info, returnValue);
    }
}

void DebugController::onInstruction(const InstructionInfo& info) {
    std::shared_ptr<lingo::vm::TraceListener> delegate;
    {
        std::lock_guard lock(mutex_);
        currentInstructionInfo_ = info;
        delegate = delegateListener_;
    }
    if (delegate) {
        delegate->onInstruction(info);
    }

    std::optional<DebugSnapshot> pausedSnapshot;
    {
        std::unique_lock lock(mutex_);
        if (checkBreakLocked(info).shouldPause) {
            state_ = DebugState::Paused;
            stepMode_ = StepMode::None;
            currentSnapshot_ = captureSnapshotLocked(info);
            pausedSnapshot = currentSnapshot_;
        }
    }
    if (!pausedSnapshot.has_value()) {
        return;
    }

    notifyPaused(*pausedSnapshot);

    std::unique_lock lock(mutex_);
    pauseCv_.wait(lock, [this] {
        return state_ != DebugState::Paused;
    });
}

bool DebugController::needsInstructionTrace() const {
    return true;
}

void DebugController::onVariableSet(std::string_view type, std::string_view name, const lingo::Datum& value) {
    std::shared_ptr<lingo::vm::TraceListener> delegate;
    {
        std::lock_guard lock(mutex_);
        delegate = delegateListener_;
    }
    if (delegate) {
        delegate->onVariableSet(type, name, value);
    }
}

void DebugController::onError(std::string_view message, std::string_view error) {
    std::shared_ptr<lingo::vm::TraceListener> delegate;
    {
        std::lock_guard lock(mutex_);
        delegate = delegateListener_;
    }
    if (delegate) {
        delegate->onError(message, error);
    }
}

void DebugController::onDebugMessage(std::string_view message) {
    std::shared_ptr<lingo::vm::TraceListener> delegate;
    {
        std::lock_guard lock(mutex_);
        delegate = delegateListener_;
    }
    if (delegate) {
        delegate->onDebugMessage(message);
    }
}

void DebugController::setGlobalsSnapshot(std::map<std::string, lingo::Datum> globals) {
    std::lock_guard lock(mutex_);
    globalsSnapshot_ = std::move(globals);
}

void DebugController::setLocalsSnapshot(std::map<std::string, lingo::Datum> locals) {
    std::lock_guard lock(mutex_);
    localsSnapshot_ = std::move(locals);
}

bool DebugController::isAwaitingStepContinuation() const {
    std::lock_guard lock(mutex_);
    return state_ == DebugState::Stepping && stepMode_ != StepMode::None;
}

void DebugController::reset() {
    {
        std::lock_guard lock(mutex_);
        state_ = DebugState::Running;
        stepMode_ = StepMode::None;
        callDepth_ = 0;
        targetCallDepth_ = 0;
        pauseRequested_ = false;
        handlerInfoStack_.clear();
        callStack_.clear();
        currentHandlerInfo_.reset();
        currentInstructionInfo_.reset();
        currentSnapshot_.reset();
    }
    pauseCv_.notify_all();
}

bool DebugController::isPaused() const {
    std::lock_guard lock(mutex_);
    return state_ == DebugState::Paused;
}

DebugState DebugController::state() const {
    std::lock_guard lock(mutex_);
    return state_;
}

StepMode DebugController::stepMode() const {
    std::lock_guard lock(mutex_);
    return stepMode_;
}

int DebugController::callDepth() const {
    std::lock_guard lock(mutex_);
    return callDepth_;
}

std::optional<std::string> DebugController::currentHandlerName() const {
    std::lock_guard lock(mutex_);
    return currentHandlerInfo_.has_value()
        ? std::optional<std::string>{currentHandlerInfo_->handlerName}
        : std::nullopt;
}

std::optional<DebugSnapshot> DebugController::currentSnapshot() const {
    std::lock_guard lock(mutex_);
    return currentSnapshot_;
}

void DebugController::stepInto() {
    bool shouldNotify = false;
    {
        std::lock_guard lock(mutex_);
        if (state_ != DebugState::Paused) {
            return;
        }
        stepMode_ = StepMode::StepInto;
        state_ = DebugState::Stepping;
        shouldNotify = true;
    }
    if (shouldNotify) {
        notifyResumed();
        pauseCv_.notify_all();
    }
}

void DebugController::stepOver() {
    bool shouldNotify = false;
    {
        std::lock_guard lock(mutex_);
        if (state_ != DebugState::Paused) {
            return;
        }
        stepMode_ = StepMode::StepOver;
        targetCallDepth_ = callDepth_;
        state_ = DebugState::Stepping;
        shouldNotify = true;
    }
    if (shouldNotify) {
        notifyResumed();
        pauseCv_.notify_all();
    }
}

void DebugController::stepOut() {
    bool shouldNotify = false;
    {
        std::lock_guard lock(mutex_);
        if (state_ != DebugState::Paused) {
            return;
        }
        stepMode_ = StepMode::StepOut;
        targetCallDepth_ = callDepth_ - 1;
        state_ = DebugState::Stepping;
        shouldNotify = true;
    }
    if (shouldNotify) {
        notifyResumed();
        pauseCv_.notify_all();
    }
}

void DebugController::continueExecution() {
    bool shouldNotify = false;
    {
        std::lock_guard lock(mutex_);
        if (state_ != DebugState::Paused) {
            return;
        }
        stepMode_ = StepMode::None;
        state_ = DebugState::Running;
        shouldNotify = true;
    }
    if (shouldNotify) {
        notifyResumed();
        pauseCv_.notify_all();
    }
}

void DebugController::pause() {
    std::lock_guard lock(mutex_);
    if (state_ == DebugState::Running || state_ == DebugState::Stepping) {
        pauseRequested_ = true;
    }
}

bool DebugController::toggleBreakpoint(int scriptId, std::string handlerName, int offset) {
    bool added = false;
    {
        std::lock_guard lock(mutex_);
        added = breakpointManager_.toggleBreakpoint(scriptId, std::move(handlerName), offset).has_value();
    }
    notifyBreakpointsChanged();
    return added;
}

Breakpoint DebugController::addBreakpoint(int scriptId, std::string handlerName, int offset) {
    Breakpoint breakpoint;
    {
        std::lock_guard lock(mutex_);
        breakpoint = breakpointManager_.addBreakpoint(scriptId, std::move(handlerName), offset);
    }
    notifyBreakpointsChanged();
    return breakpoint;
}

std::optional<Breakpoint> DebugController::removeBreakpoint(int scriptId,
                                                            const std::string& handlerName,
                                                            int offset) {
    std::optional<Breakpoint> removed;
    {
        std::lock_guard lock(mutex_);
        removed = breakpointManager_.removeBreakpoint(scriptId, handlerName, offset);
    }
    if (removed.has_value()) {
        notifyBreakpointsChanged();
    }
    return removed;
}

std::optional<Breakpoint> DebugController::toggleBreakpointEnabled(int scriptId,
                                                                   const std::string& handlerName,
                                                                   int offset) {
    std::optional<Breakpoint> breakpoint;
    {
        std::lock_guard lock(mutex_);
        breakpoint = breakpointManager_.toggleEnabled(scriptId, handlerName, offset);
    }
    if (breakpoint.has_value()) {
        notifyBreakpointsChanged();
    }
    return breakpoint;
}

void DebugController::setBreakpoint(Breakpoint breakpoint) {
    {
        std::lock_guard lock(mutex_);
        breakpointManager_.setBreakpoint(std::move(breakpoint));
    }
    notifyBreakpointsChanged();
}

void DebugController::setBreakpoints(std::map<int, std::set<int>> breakpoints) {
    {
        std::lock_guard lock(mutex_);
        breakpointManager_.setFromOffsetMap(breakpoints);
    }
    notifyBreakpointsChanged();
}

std::map<int, std::set<int>> DebugController::breakpoints() const {
    std::lock_guard lock(mutex_);
    return breakpointManager_.toOffsetMap();
}

void DebugController::clearAllBreakpoints() {
    {
        std::lock_guard lock(mutex_);
        breakpointManager_.clearAll();
    }
    notifyBreakpointsChanged();
}

bool DebugController::hasBreakpoint(int scriptId, const std::string& handlerName, int offset) const {
    std::lock_guard lock(mutex_);
    return breakpointManager_.hasBreakpoint(scriptId, handlerName, offset);
}

std::optional<Breakpoint> DebugController::getBreakpoint(int scriptId,
                                                         const std::string& handlerName,
                                                         int offset) const {
    std::lock_guard lock(mutex_);
    return breakpointManager_.getBreakpoint(scriptId, handlerName, offset);
}

BreakpointManager& DebugController::breakpointManager() {
    return breakpointManager_;
}

const BreakpointManager& DebugController::breakpointManager() const {
    return breakpointManager_;
}

std::string DebugController::serializeBreakpoints() const {
    std::lock_guard lock(mutex_);
    return breakpointManager_.serialize();
}

void DebugController::deserializeBreakpoints(std::string data) {
    {
        std::lock_guard lock(mutex_);
        breakpointManager_.deserialize(std::move(data));
    }
    notifyBreakpointsChanged();
}

WatchExpression DebugController::addWatchExpression(std::string expression) {
    WatchExpression watch;
    {
        std::lock_guard lock(mutex_);
        watch = WatchExpression::create(std::move(expression));
        watchExpressions_.push_back(watch);
    }
    notifyWatchExpressionsChanged();
    return watch;
}

bool DebugController::removeWatchExpression(const std::string& id) {
    bool removed = false;
    {
        std::lock_guard lock(mutex_);
        const auto before = watchExpressions_.size();
        watchExpressions_.erase(std::remove_if(watchExpressions_.begin(),
                                               watchExpressions_.end(),
                                               [&id](const WatchExpression& watch) {
                                                   return watch.id == id;
                                               }),
                                watchExpressions_.end());
        removed = watchExpressions_.size() != before;
    }
    if (removed) {
        notifyWatchExpressionsChanged();
    }
    return removed;
}

std::optional<WatchExpression> DebugController::updateWatchExpression(const std::string& id,
                                                                      std::string expression) {
    std::optional<WatchExpression> updated;
    {
        std::lock_guard lock(mutex_);
        for (auto& watch : watchExpressions_) {
            if (watch.id == id) {
                watch = watch.withExpression(std::move(expression));
                updated = watch;
                break;
            }
        }
    }
    if (updated.has_value()) {
        notifyWatchExpressionsChanged();
    }
    return updated;
}

std::vector<WatchExpression> DebugController::watchExpressions() const {
    std::lock_guard lock(mutex_);
    return watchExpressions_;
}

std::vector<WatchExpression> DebugController::evaluateWatchExpressions() {
    std::lock_guard lock(mutex_);
    return evaluateWatchExpressionsLocked();
}

void DebugController::clearWatchExpressions() {
    {
        std::lock_guard lock(mutex_);
        watchExpressions_.clear();
    }
    notifyWatchExpressionsChanged();
}

std::vector<CallFrame> DebugController::callStackSnapshot() const {
    std::lock_guard lock(mutex_);
    return callStack_;
}

DebugController::BreakResult DebugController::checkBreakLocked(const InstructionInfo& info) {
    if (pauseRequested_) {
        pauseRequested_ = false;
        return BreakResult{true};
    }

    const bool suppressBreakpoints =
        (stepMode_ == StepMode::StepOver || stepMode_ == StepMode::StepOut) &&
        callDepth_ > targetCallDepth_;
    if (!suppressBreakpoints && currentHandlerInfo_.has_value()) {
        const auto breakpoint = breakpointManager_.getBreakpoint(
            currentHandlerInfo_->scriptId,
            currentHandlerInfo_->handlerName,
            info.offset);
        if (breakpoint.has_value() && breakpoint->enabled) {
            return BreakResult{true};
        }
    }

    bool shouldStep = false;
    switch (stepMode_) {
        case StepMode::StepInto:
            shouldStep = true;
            break;
        case StepMode::StepOver:
            shouldStep = callDepth_ <= targetCallDepth_;
            break;
        case StepMode::StepOut:
        case StepMode::None:
            shouldStep = false;
            break;
    }
    return BreakResult{shouldStep};
}

DebugSnapshot DebugController::captureSnapshotLocked(const InstructionInfo& info) {
    const auto handlerInfo = currentHandlerInfo_.value_or(HandlerInfo{});
    const auto locals = !info.localsSnapshot.empty() ? toOrderedMap(info.localsSnapshot) : localsSnapshot_;
    const auto globals = !info.globalsSnapshot.empty() ? toOrderedMap(info.globalsSnapshot) : globalsSnapshot_;
    localsSnapshot_ = locals;
    globalsSnapshot_ = globals;
    const auto watches = evaluateWatchExpressionsLocked();

    return DebugSnapshot{
        handlerInfo.scriptId,
        handlerInfo.scriptDisplayName,
        handlerInfo.handlerName,
        info.offset,
        info.bytecodeIndex,
        info.opcode,
        info.argument,
        info.annotation,
        {},
        info.stackSnapshot,
        locals,
        globals,
        handlerInfo.arguments,
        handlerInfo.receiver.isVoid() ? std::optional<lingo::Datum>{} : std::optional<lingo::Datum>{handlerInfo.receiver},
        callStack_,
        watches,
    };
}

ExpressionEvaluator::EvaluationContext DebugController::buildEvaluationContextLocked() const {
    ExpressionEvaluator::EvaluationContext context;
    context.locals = localsSnapshot_;
    context.globals = globalsSnapshot_;
    if (currentHandlerInfo_.has_value()) {
        for (std::size_t index = 0; index < currentHandlerInfo_->arguments.size(); ++index) {
            context.params["arg" + std::to_string(index)] = currentHandlerInfo_->arguments[index];
        }
        if (!currentHandlerInfo_->receiver.isVoid()) {
            context.receiver = currentHandlerInfo_->receiver;
        }
    }
    return context;
}

std::vector<WatchExpression> DebugController::evaluateWatchExpressionsLocked() {
    const auto context = buildEvaluationContextLocked();
    std::vector<WatchExpression> evaluated;
    evaluated.reserve(watchExpressions_.size());
    for (const auto& watch : watchExpressions_) {
        const auto result = expressionEvaluator_.evaluate(watch.expression, context);
        evaluated.push_back(result.succeeded()
                                ? watch.withValue(*result.value)
                                : watch.withError(result.error.value_or("Evaluation failed")));
    }
    return evaluated;
}

void DebugController::notifyPaused(const DebugSnapshot& snapshot) {
    std::vector<DebugStateListener*> listeners;
    {
        std::lock_guard lock(mutex_);
        listeners = listenersSnapshotLocked();
    }
    for (auto* listener : listeners) {
        listener->onPaused(snapshot);
    }
}

void DebugController::notifyResumed() {
    std::vector<DebugStateListener*> listeners;
    {
        std::lock_guard lock(mutex_);
        listeners = listenersSnapshotLocked();
    }
    for (auto* listener : listeners) {
        listener->onResumed();
    }
}

void DebugController::notifyBreakpointsChanged() {
    std::vector<DebugStateListener*> listeners;
    {
        std::lock_guard lock(mutex_);
        listeners = listenersSnapshotLocked();
    }
    for (auto* listener : listeners) {
        listener->onBreakpointsChanged();
    }
}

void DebugController::notifyWatchExpressionsChanged() {
    std::vector<DebugStateListener*> listeners;
    {
        std::lock_guard lock(mutex_);
        listeners = listenersSnapshotLocked();
    }
    for (auto* listener : listeners) {
        listener->onWatchExpressionsChanged();
    }
}

std::vector<DebugStateListener*> DebugController::listenersSnapshotLocked() const {
    return listeners_;
}

std::map<std::string, lingo::Datum> DebugController::toOrderedMap(
    const std::unordered_map<std::string, lingo::Datum>& values) {
    std::map<std::string, lingo::Datum> ordered;
    for (const auto& [key, value] : values) {
        ordered[key] = value;
    }
    return ordered;
}

} // namespace libreshockwave::player::debug
